File: 01_introduction_to_algorithms/ruby/01_binary_search.rb
def binary_search(list, item)
  low = 0
  high = list.length - 1

  while low <= high
    mid = (low + high) / 2
    guess = list[mid]

    if guess == item
      return mid
    elsif guess > item
      high = mid - 1
    else
      low = mid + 1
    end
  end

  return nil
end

my_list = [1, 3, 5, 7, 9]
puts binary_search(my_list, 3) # => 1

# We need to use .inspect here because just printing nil
# would output an empty string
puts binary_search(my_list, -1).inspect # => nil

File: 01_introduction_to_algorithms/perl5/01_binary_search.pl
sub binary_search {

    my ( $item, @list ) = @_;

    # The borders inside where we are searching in:
    my $low  = 0;
    my $high = $#list;

    # Do that until found or until just one element remains:
    while ( $low <= $high ) {

        # Detect the middle of the array:
        my $mid = int( ( $low + $high ) / 2 );
        my $guess = $list[$mid];

        # We found it out:
        if ( $guess == $item ) {
            return $mid;
        }

        # The guess is too high, decrease it:
        elsif ( $guess > $item ) {
            $high = --$mid;
        }

        # The guess is too low, increase it:
        else {
            $low = ++$mid;
        }
    }

    # Nothing at all found:
    return 'None';

}

my @my_list = qw /1 3 5 7 9/;

# Should display "1":
print binary_search( 3, @my_list ), "\n";

# Should display "None":
print binary_search( -1, @my_list ), "\n";
